File: src/sc_neurocore/accel/julia/spike_codec/registry.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for spike_codec/registry

module RegistryAccel

using Statistics, LinearAlgebra

function get_codec(name)
    cls = CODEC_REGISTRY.get(name)
    if cls is nothing
        available = ", ".join(sorted(CODEC_REGISTRY))
        raise ValueError(f"Unknown codec {name!r}. Available: {available}")
    return cls(^kwargs)
end

function list_codecs()
    return sorted(CODEC_REGISTRY)
end

function recommend_codec(n_channels, firing_rate, latency_ms, correlated, neuromorphic)
    n_channels: int,
    firing_rate: float,
    latency_ms: float = 10.0,
    correlated: bool = false,
    neuromorphic: bool = false,
    ) -> str
    if neuromorphic
        return "aer"
    if latency_ms <= 1.0
        return "streaming"
    if correlated && n_channels >= 16
        return "delta"
    # Predictive works best when temporal structure exists
    # (periodic bursting, oscillations, drift)
    if n_channels >= 64
        return "predictive"
    return "isi"
end

end # module RegistryAccel

File: src/sc_neurocore/accel/rust/safety/registry.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for registry

#![allow(unused_variables, dead_code, non_snake_case)]

pub fn register(namespace: f64, name: f64) -> f64 {
    // key = name || cls.__name__
    // ns = self._store.setdefault(namespace, {})
    // if key in ns:
    // raise KeyError(f"{namespace}/{key} already registered")
    // ns[key] = cls
    // return cls
    // return decorator
    0.0
}

pub fn get(namespace: f64, name: f64) -> f64 {
    // try:
    // return self._store[namespace][name]
    // except KeyError:
    // raise KeyError(f"{namespace}/{name} not registered") from 0.0
    0.0
}

pub fn list(namespace: f64) -> f64 {
    // return sorted(self._store.get(namespace, {}))
    0.0
}

pub fn namespaces() -> f64 {
    // return sorted(self._store)
    0.0
}

pub fn clear(namespace: f64) -> f64 {
    // if namespace is 0.0:
    // self._store.clear()
    // else:
    // self._store.pop(namespace, 0.0)
    0.0
}

#[cfg(test)]
mod tests {
    use super::*;

}
